### ISSUE TITLE
I need some help

### ISSUE THREAD

**Issue by desarrollo03TR** [xs-admin/Excess#3]
Hi! your project looks great, but I need your help because I don't know how to use the vs extension.

After install the vs extensions I created a new project, then, I'm trying following this:

1- Add a .xs file to your project. 
2- Install extensions as Nuget packages. 
3- Include the extensions you need with using xs.

The step 1 is done, no problem, but I don't know which nuget packages I need to install. I Found the Excess.Extensions, the Excess.Runtime, Excess.Compiler and another packages.

Thanks!

**Comment by xs-admin** [comment]
Hello,

Some of the packages you mention are not extensions but the libs you'd need
to build your own languages. The rest are indeed extensions:

- Excess.Server <https://www.nuget.org/packages/Excess.Server/0.48.16-alpha> is
our web server extension
- Excess.Extensions
<https://www.nuget.org/packages/Excess.Extensions/0.48.10-alpha>  is a
collection of extensions you can find in the project with the same name
- Excess.Concurrent
<https://www.nuget.org/packages/Excess.Concurrent/0.48.10-alpha> is our
concurrency extension
- Excess.Extensions.NInjector
<https://www.nuget.org/packages/Excess.Extensions.NInjector/0.48.10-alpha> is
an implementation of the *inject *idiom based on the NInjector library
- Excess.Extensions.Dapper
<https://www.nuget.org/packages/Excess.Extensions.Dapper/0.48.10-alpha> is
a sql extension internally implemented with Dapper.

I hope this is useful, please feel free to ask any questions, in particular
if you are interested in creating languages.

Best Regards,
Emilio Santos

**Comment by desarrollo03TR** [comment]
Then, I need to add ".xs" files with my own languages? I this related with the code in the http://xslang.azurewebsites.net website?

**Comment by xs-admin** [comment]
No, your languages are written in C# as nay of the extensions I listed
before. You'd be using the Excess.Compiler library and using one (or many)
of our compiler helpers. When you are done, you can publish it as a nuget
packages for anyone to use,
If you give me more info about the language you intend to create I could
likely point you to something similar already in existence.

http://xslang.azurewebsites.net is a basic web compiler that internally
uses excess, but it is not an extension language itself.

**Comment by desarrollo03TR** [comment]
Thanks! I don't know how far I can go with language extensions, what I need is to extend c# language because I have several functions that I use on all my programs, for example, an "if" replacement as expression and not as statement, instead writing:

    var x = 0;
    if (!something)
        x = 5;
    else
        x = 10;

I write:

    var x = If(!something, 5, 10);

And would be great if I can write something like:

    var x = if (!something) then 5 else 10;

Or similar, or convert the switch statement in an expression match:

    var x = 
        match(month)
        { 
            case 0: "January";
            case 1: "February";
            ...
            default: "Not a month";
        }

Transformed to:

    var x = Function(() => 
    {
        switch(month)
        {
            case 0: return "January";
            ...
        }
        return "Not a month";
    }

Basically I need to replace some custom statement with my functions, "syntactic sugar", but I don't know if it is possible.

**Comment by xs-admin** [comment]
This is definitively possible, your first grammar If(condition, then, else)
is very easy, you'd register a syntax match where you watch for invocations
where the function name is "If" and then you replace that expression with a
proper if statement.

The second If() then ... else ...; is also possible, though I'm not sure it
is advisable, in this case you'd write a lexical extension (that deals with
tokens instead of AST nodes) that would look something like:

            lexical
                .match()
                    .token("If")
                    .enclosed('(', ')', named: "condition")
                    .token('then')
                    .until("else", named: "then_expression")
                    .until(';', named: "else_expression")
                        .then(YourFunction)

This of course you have to play with, but the gist is: in "YourFunction"
you will receive the results of parsing (including  condition,
then_expression, else_expression) and you will build a Roslyn IfStatement
with it.

Hope it helps.

**Comment by desarrollo03TR** [comment]
Thanks. Can you share an project example?

**Comment by xs-admin** [comment]
You could look at the "function" implementation here:

https://github.com/xs-admin/Excess/blob/master/Excess.Compiler/xslang/Functions.cs

And, in reality, all the files in  the folder

https://github.com/xs-admin/Excess/blob/master/Excess.Compiler/xslang

I recommend you use tests like this one to test your language:

https://github.com/xs-admin/Excess/blob/master/Excess.Compiler/Tests/FunctionTests.cs

Hope it helps!

**Comment by desarrollo03TR** [comment]
Thanks a lot! I'll try to develop my own library.

**Comment by xs-admin** [comment]
Please let me know how it goes

**Comment by desarrollo03TR** [comment]
Another question: if I want to use the dapper extensions (or another extensions), how can I use it? Is it possible to use it?

EDIT: after some testing, I can use the arrays, match and function extensions, but contract and typedef are not working (or I don't know how to use it).

**Comment by xs-admin** [comment]
you use it by including it in your .xs file:

using xs.extension_name;

**Comment by desarrollo03TR** [comment]
Thanks, after some testing, everything except typedef is working.

**Comment by xs-admin** [comment]
Yes, in retrospect I should remove the typedef, there is not enough
advantages to keep it.

**Comment by desarrollo03TR** [comment]
typedef could be useful if it create a really new type, for example, a type definition like this:

    typedef byte Age;

could be transpiled to:

    public partial struct Age : IEquatable<Age>, IEquatable<byte>, IComparable<byte>, IComparable<Age>, IConvertible, IFormattable
    {
        private readonly byte value;
        public  byte Value { get { return value; }}
        public Age(byte value)
        {
            this.value = value;
        }
        
        // all the equality and comparision methods
    }

A great option for value types creation.

**Comment by xs-admin** [comment]
That is a very good point, when initially implemented I was playing with
extension cooperation (multiple .xs files sharing the typedefs) but I have
desisted on that idea so far.
Things get very complicated in those scenarios.
